yksmp/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+            let dwreg = self.read_u16();
+            self.read_u16();
+
+            let location = match kind {
+                0x01 => {
+                    self.read_i32();
+                    Location::Register(dwreg)
+                }
+                0x02 => {
+                    let offset = self.read_i32();
+                    Location::Direct(dwreg, offset)
+                }
+                0x03 => {
+                    let offset = self.read_i32();
+                    Location::Indirect(dwreg, offset)
+                }
+                0x04 => {
+                    let offset = self.read_u32();
+                    Location::Constant(offset)
+                }
+                0x05 => {
+                    let offset = self.read_i32();
+                    Location::LargeConstant(consts[usize::try_from(offset).unwrap()])
+                }
+                _ => unreachable!(),
+            };
+
+            v.push(location)
+        }
+        v
+    }
+
+    fn read_liveouts(&mut self, num: u16) -> Vec<LiveOut> {
+        let mut v = Vec::new();
+        for _ in 0..num {
+            let dwreg = self.read_u16();
+            let size = self.read_u8();
+            v.push(LiveOut { dwreg, size });
+        }
+        v
+    }
+
+    fn align_8(&mut self) {
+        self.offset += (8 - (self.offset % 8)) % 8;
+    }
+
+    fn read_u8(&mut self) -> u8 {
+        let d = u8::from_ne_bytes(self.data[self.offset..self.offset + 1].try_into().unwrap());
+        self.offset += 1;
+        d
+    }
+
+    fn read_u16(&mut self) -> u16 {
+        let d = u16::from_ne_bytes(self.data[self.offset..self.offset + 2].try_into().unwrap());
+        self.offset += 2;
+        d
+    }
+
+    fn read_u32(&mut self) -> u32 {
+        let d = u32::from_ne_bytes(self.data[self.offset..self.offset + 4].try_into().unwrap());
+        self.offset += 4;
+        d
+    }
+
+    fn read_i32(&mut self) -> i32 {
+        let d = i32::from_ne_bytes(self.data[self.offset..self.offset + 4].try_into().unwrap());
+        self.offset += 4;
+        d
+    }
+
+    fn read_u64(&mut self) -> u64 {
+        let d = u64::from_ne_bytes(self.data[self.offset..self.offset + 8].try_into().unwrap());
+        self.offset += 8;
+        d
+    }
+}
+
+/// Maps stackmap records to the addresses where they are relevant. Allows users to retrieve the
+/// appropriate record during deoptimisation.
+struct SMQuery {
+    map: HashMap<u64, Record>,
+}
+
+impl SMQuery {
+    #[allow(dead_code)]
+    fn get_record(&self, addr: u64) -> Option<&Record> {
+        self.map.get(&addr)
+    }
+
+    #[allow(dead_code)]
+    fn get_liveout(&self, addr: u64) -> Option<&Vec<LiveOut>> {
+        self.map.get(&addr).map(|r| &r.liveouts)
+    }
+
+    /// Get all locations given the return address of a __llvm_deoptimize function.
+    fn get_locations(&self, addr: u64) -> Option<&Vec<Location>> {
+        self.map.get(&addr).map(|r| &r.locs)
+    }
+}
+
+struct Registers {
```

Comment:
These registers are x86_64 specific. We discussed potentially guarding this with a `cfg`? Is it worth it?

Reply:
I actually think these should just be `u64` instead. There's a comment at the top saying this whole parser is x64 specific and we'll need to either make it dynamic or write new ones, once LLVM adds support for other architectures.

Reply:
ok.